rohmu/object_storage/s3.py

```diff
@@ -592,6 +596,93 @@ def check_or_create_bucket(self) -> None:
             self.stats.operation(StorageOperation.create_bucket)
             self.s3_client.create_bucket(**args)
 
+    def create_concurrent_upload(
+        self, key: str, metadata: Optional[Metadata] = None, mimetype: Optional[str] = None
+    ) -> ConcurrentUploadId:
+        args: dict[str, Any] = {
+            "Bucket": self.bucket_name,
+            "Key": self.format_key_for_backend(key, remove_slash_prefix=True),
+        }
+        if metadata:
+            metadata = self.sanitize_metadata(metadata)
+            args["Metadata"] = metadata
+        if self.encrypted:
+            args["ServerSideEncryption"] = "AES256"
+        if mimetype:
+            args["ContentType"] = mimetype
+
+        self.stats.operation(StorageOperation.create_multipart_upload)
+        try:
+            cmu_response = self.s3_client.create_multipart_upload(**args)
+        except botocore.exceptions.ClientError as ex:
+            raise StorageError("Failed to initiate multipart upload for {}".format(key)) from ex
+
+        upload = ConcurrentUploadData("aws", cmu_response["UploadId"], key)
+        upload_id = self._new_concurrent_upload(upload, metadata)
+        return upload_id
+
+    def complete_concurrent_upload(self, upload_id: ConcurrentUploadId) -> None:
+        concurrent_data, metadata, chunks = self._get_concurrent_upload(upload_id)
+        backend_key = self.format_key_for_backend(concurrent_data.key, remove_slash_prefix=True)
+        sorted_chunks = sorted(
+            ({"ETag": etag, "PartNumber": number} for number, etag in chunks.items()),
+            key=lambda part: cast(int, part["PartNumber"]),
+        )
+        try:
+            self.s3_client.complete_multipart_upload(
+                Bucket=self.bucket_name,
+                Key=backend_key,
+                MultipartUpload={"Parts": sorted_chunks},  # type: ignore[typeddict-item]
+                UploadId=concurrent_data.backend_id,
+                RequestPayer="requester",
+            )
+            del self._concurrent_uploads[upload_id]
+            # NOTE: metadata is not available if we are resuming. size is never available.
+            self.notifier.object_created(key=backend_key, size=None, metadata=metadata)
```

Comment:
why is notifying useless? Can you elaborate?

Reply:
the only consumer currently is robjecter which drop the notification if `size` is none. plus if metadata is also `None` then robjecter is give wrong information

Reply:
I'm thinking that, since we are not handling resume, we could somehow keep track of the size of the chunks that are uploaded and provide correct size... but if we ever need to add support for resume this adds an additional requirement regarding the state of the upload.

Reply:
I noticed that the same things happen, for example, in the `copy_object` method.

I'd leave this code as is, and hopefully in the future we could be able to provide a value for `size`... S3 list-parts has information on the size of each part, so it would be possible to obtain it...

Reply:
please drop this, a wrong metadata is completely bonkers, the copy_file thing is bug waiting to happen. But during last conversation on copy_file the argument was `copy_file` is almost never used.